netket/graph/grid.py

```diff
@@ -0,0 +1,55 @@
+from graph import Graph
+
+import numpy as _np
+import networkx as _nx
+
+
+class HyperGrid(Graph):
+    r"""A hypercube lattice of side L in d dimensions. 
+	Periodic boundary conditions can also be imposed"""
+
+    def __init__(self, length, pbc=True):
+        """
+		Constructs a new ``Grid`` given its length vector.
+
+		Args:
+			length: Side length of the hypergrid.
+				It must always be >= 1,
+				but if ``pbc==True`` then the minimal valid length is 3.
+				It can also be a list[int] where each entry is the size of a dimension.
+			n_dim: Dimension of the hypercube. It must be at least 1.
+			pbc: If ``True```then the constructed hypercube will have
+				periodic boundary conditions, otherwise open boundary conditions
+				are imposed.
+		Examples:
+		    A 10x10 square lattice with periodic boundary conditions can be
+		    constructed as follows:
+
+		    >>> import netket
+		    >>> g=netket.graph.PyHypercube(length=10,n_dim=2,pbc=True)
+		    >>> print(g.n_sites)
+		    100
+
+		    Also, a 2x2x3 lattice with open boundary conditions can be constructed as follows:
+		    >>> g=netket.graph.PyHypercube(length=[2,2,3], pbc=False)
+		    >>> print(g.n_sites)
+		    12
+		"""
+
+        assert isinstance(length, int) or isinstance(length, list)
+        assert isinstance(n_dim, int) and n_dim >= 1
+
+        if isinstance(length, int):
+            length = [length] * n_dim
+        else:
+            try:
+                condition = [isinstance(x, int) and x >= 1 for x in length]
+                if sum(condition) != len(length):
+                    raise TypeError
+            except TypeError:
+                raise TypeError("Components of length must be integers greater than 1")
+
+        graph = _nx.generators.lattice.grid_graph(length, periodic=pbc)
+        newnames = {old: new for new, old in enumerate(graph.nodes)}
+        graph = _nx.relabel_nodes(graph, newnames)
+        super().__init__(list(graph.edges))
```

Comment:
This process seems to change the order of vertices again after you rename them, when initializing `Graph` in the base class constructor with just the edge list, as done in line 55:
```python
>>> grid = nk.graph.PyHypercube(4, 1, pbc=True)
>>> grid.graph.nodes
NodeView((0, 1, 3, 2))
>>> grid.automorphisms
[[0, 1, 3, 2],
 [0, 3, 1, 2],
 [1, 0, 2, 3],
 [3, 0, 2, 1],
 [2, 1, 3, 0],
 [2, 3, 1, 0],
 [1, 2, 0, 3],
 [3, 2, 0, 1]]
```
Note that the identity permutation is not in the list (or, actually, it is because its `[0, 1, 3, 2]`, but that's pretty counter-intuitive).

I think it would be better, to initialize the `Graph` directly with the networx graph created by of `grid_graph`. That way, there is no reordering of nodes.

Reply:
The thing is that `grid_graph` creates hypergrids where nodes are not integers but tuples, which is why I renamed the nodes.

Reply:
I think that if we want to stick to the old automorphisms we need to input the edges by hand, following the same convention used in old ```Hypercube``` or, we can just stick with this strange ordering that ```grid_graph``` gives... but it will require some reshuffling of the autormophism to make sure that it is correctly sorted, as @femtobit pointed out.

Reply:
In practice, you'd need to first sort the dict given by the networkx function with something like
```python
isos=nx.isomorphism.GraphMatcher(grid.graph, grid.graph).isomorphisms_iter()

for iso in isos:
      sorted_d = sorted(iso.items(), key=lambda x: x[1]
      iso = sorted_d
```

to be checked it this actually works... !

Reply:
I found that the creation of the graph is a bit weird for the case pointed out by @femtobit. We can create an hypergrid (in this case a 1D chain) like this:

```
nx.generators.lattice.grid_graph([4], periodic=True).nodes             
Out[26]: NodeView((0, 1, 2, 3))
```
Here, the nodes are nicely ordered, and the edges are given by:
```
nx.generators.lattice.grid_graph([4], periodic=True).edges             
Out[27]: EdgeView([(0, 1), (0, 3), (1, 2), (2, 3)])
```
Now, this graph is created at `hypergrid.py`, and then a list of its edges is passed to the general-purpose Graph in `graph.py`. But this messes the order of the nodes:
```
nx.MultiGraph(nx.generators.lattice.grid_graph([4], periodic=True).edges).nodes                                                               
Out[30]: NodeView((0, 1, 3, 2))
```
The same happens if one uses `nx.Graph` instead of `nx.MultiGraph`. Basically, if one passes the list of edges `[(0, 1), (0, 3), (1, 2), (2, 3)]` one gets the ordering problem, but if one passes a list of edges `[(0, 1), (1, 2), (2, 3), (3, 0)]`, then the ordering of the nodes is correct. In both cases the edges are the same.

Reply:
I found that the only solution was to pass to the graph creator first the nodes, and then the edges, so that the nodes are created first and in an ordered fashion. I just pushed the changes.

Reply:
Sounds good.